canonical/ubuntu-advantage-client
# apt-hook: print static messages on apt upgrade/dist-upgrade

## New behavior

* on `apt upgrade/dist-upgrade`
  * find esm packages (names and count)
  * will insert values into message template for contract-expiry message if the template exists and print the message
* when called as `hook process-templates`
  * find esm packages (names and count)
  * will insert values into hardcoded message templates that exist
  * outputs those messages to static files with the values filled in
* on `apt update`
  * updates all the template files, but no longer prints the messages it used to


## Proposed Commit Message


> apt-hook: process template files and print messages with esm stats
>
> apt-esm-hook will now process a set of .tmpl files in the ua messages
> directory, both when apt calls it and when called separately with the
> "process-templates" argument. It replaces variables in the tmpl files
> with the values of ESM Apps/Infra package names and count. It then
> writes the result of the replacement back to the ua messages
> directory.
>
> When called on Post-Invoke for upgrade/dist-upgrade, the hook will
> print out the contents of message files if they are present.
>
> Fixes: #1546 

## Test Steps


In a new xenial lxd container, preferably one with outstanding esm updates:
1. `mkdir -p /var/lib/ubuntu-advantage/messages`
2. `echo "esm-apps {ESM_APPS_PKG_COUNT} {ESM_APPS_PACKAGES}; esm-infra {ESM_INFRA_PKG_COUNT} {ESM_INFRA_PACKAGES}" > /var/lib/ubuntu-advantage/messages/contract-expired-apt-upgrade.tmpl`
3. `/usr/lib/ubuntu-advantage/apt-esm-hook process-templates`
4. `cat /var/lib/ubuntu-advantage/messages/contract-expired-apt-upgrade` and make sure it has the expected output
5. `apt upgrade` and ensure expected output appears


## Desired commit type::


## Checklist:
